### ISSUE TITLE
Refactor `emitter::instrDesc` and supporting code for XARCH

### ISSUE THREAD

**Issue by CarolEidt** [dotnet/coreclr#14523]
As new instructions have been added for SSE and AVX, the existing encoding tables aren't adequately expressive, resulting in a great deal of conditional code that impacts throughput. This should be refactored both for throughput, as well as to avoid complex conditional code that is harder to maintain.
This will contribute to the resolution of #7454.

**Comment by fiigii** [comment]
@CarolEidt Thank you for opening this issue. If nobody is working on this, please assign to me. I think implementing the new AVX/SSE hardware intrinsics is a good opportunity to investigate and refactor the instruction encoding tables.

**Comment by mikedn** [comment]
I looked a bit at this back when adding `INS_bt` and it looks like one possible solution is to extend instruction formats so that each format maps to a emission code sequence that needs no or little conditional code. Mapping from the instruction format to the emission code is already done using a `switch` that produces a jump table so adding more cases to it should not hurt.

Then the instruction table would contain some kind of "base" format to which update modes are added.

Something like shift formats (`IF_RRW_SHF`, `IF_MRW_SHF` etc.) except that the format would be taken from the table instead of being another piece of conditional code such as `emitter::emitMapFmtForIns`.

Remains to be seen if such an approach is sufficient, the x86 instruction encoding is very complex. But then, it may be enough to ensure that common instructions are handled efficiently. As is now adding new instructions, that perhaps aren't used often, risks affecting the performance of existing code paths.

**Comment by fiigii** [comment]
@mikedn This is a good idea, but the current format mechanism is already so complex. Shall we add a new "format" type rather than extend the current mechanism?

**Comment by mikedn** [comment]
I can only guess that adding a new format would be worse from a performance perspective. But it still should be better than what we have now.

Even if we only get rid of `IsSSEOrAVXInstruction` & co. we still might get an improvement.

**Comment by mikedn** [comment]
Speaking of `IsSSEOrAVXInstruction`, I recently tried to add SHLX/SHRX/SARX and ran into some problems because they use VEX encoding but they aren't exactly AVX instructions...

**Comment by CarolEidt** [comment]
> Shall we add a new "format" type rather than extend the current mechanism?

I'm not sure what you mean by a new "format" type - would that be like a secondary format?
If so, I would agree with @mikedn that adding formats would probably be a better approach - and I would hope that the current complexity of the format mechanism would  actually be ameliorated by eliminating some of the conditional handling of the different cases covered by a single format.
That said, I can't say that I have spent enough time (especially recently) analyzing this code to say definitively.

**Comment by fiigii** [comment]
Besides `IsSSEOrAVXInstruction`, we should encode certain REX/VEX prefix bits, VEX non-destructive source (NDS) info, instruction size, and maybe more.  Encoding VEX.NDS can get rid of `IsDstSrcSrcAVXInstruction`.

**Comment by fiigii** [comment]
> If so, I would agree with @mikedn that adding formats would probably be a better approach - and I would hope that the current complexity of the format mechanism would actually be ameliorated by eliminating some of the conditional handling of the different cases covered by a single format.

Yes. It would be really good to ameliorate complexity of the format mechanism. Let me investigate more emitter code and instruction encoding rules.

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib

**Comment by tannergooding** [comment]
I missed that there was an existing issue for this and opened https://github.com/dotnet/coreclr/issues/16333 over the weekend. I've copied the relevant bits below.

**Comment by tannergooding** [comment]
I believe the two primary limitations in the existing setup are:

A specific ordering of definitions is required
This limits the ability to easily refactor or re-order the existing data
The current definitions assume all encodings of an instruction share the same "properties"
Some encodings belong to different ISAs
Some encodings support different architectures (32-bit vs 64-bit)
Some encodings read/write flags, while others may not
Some encodings may vary on the prefixes supported/required
etc...

**Comment by tannergooding** [comment]
I believe the first issue (the current implementation requires definitions to be particularly ordered) is more trivial to resolve (and I think I have such a refactoring complete locally, at the very least all the tests are passing).

It looks like we currently require the entries to be defined in the following order:
* INST5
* INST4
* INST3
* INST2
* INST1
* INST0

It looks like this was done in order to save space in the various `insCodes` tables. However, this quickly broke down with the introduction of the SIMD instructions where they all use the `INST3` definitions and generally have two `BAD_CODE` encodings.

Updating the tables to define an entry for all instructions removes most of the ordering limitations. The remaining ordering limitations are with certain instruction "groups" (such as `imul_AX` through `imul_R15`, or the `JCC`, `SETCC`, or `MOVCC` instructions). However these remaining limitations cause less issues.

**Comment by tannergooding** [comment]
The second issue (flags and other metadata are on a per instruction, rather than a per encoding basis) is more complicated.

I believe we would need to update the metadata to define an instruction (id + name). Each instruction would then have multiple encodings. Each encoding would then define the following data:
* ISA
* Base Encoding (the actual bytes for the encoding)
* CPU Flags Read
  * A general bool is defined in the current table, but looks unused
  * This would actually be a bitmap of the individual flags read
* CPU Flags Written
  * A general bool is defined in the current table, but looks unused
  * This would actually be a bitmap of the individual flags written
* Encoding Flags
  * Supports 32-bit
  * Supports 64-bit
  * Supports 2-byte VEX Prefix
  * Supports 3-byte VEX Prefix
  * etc

How these encodings are defined per instruction is probably worth some discussion. The current Architecture manual specifies there are a few different instruction formats (see Volume 2, Chapter 2):
* Protected Mode, Real-Address Mode, and Virtual 8086 Mode Encoding
* IA-32E Mode Encoding
* AVX Encoding (VEX Encoding)
* VEX Encoding for GPR
* AVX-512 Encoding (EVEX Encoding)

In order to limit the amount of table data that is unused and to support future encodings, it would likely be beneficial to construct some mechanism for defining encodings in each of these categories, but not requiring each category per instruction. This will likely also simplify the various checks/code-paths required in the emitter.

**Comment by CarolEidt** [comment]
Another factor to consider is how to deal with the instruction "classes" that are treated like a single instruction by the code generator, and ideally would continue to be so, but that have different behaviors that need to be handled for the various ISAs (e.g. SSE vs. AVX), as well as when they have a memory operand. The current `emitter::IsDstDstSrcAVXInstruction()` and `emitter::IsDstSrcSrcAVXInstruction()` methods address some of this, but imperfectly, e.g. even if we made this property explicit, it wouldn't handle the memory operand case as illustrated in #17560.

**Comment by mikedn** [comment]
Some observations about the current instruction table:
* `sahf` is not used (leftover from x87 FP)
* `loop` is not used. Was it ever used? Usually it's not recommended, it can have lower performance than normal branch instructions.
* `push_hide` and `pop_hide` are kind of weird, perhaps the same information can be communicated by other means? But then they're only 2 instructions among hundreds so it probably does not matter.
* The need for `imul_reg` instructions is probably due to codegen - emitter interfacing problems. If you need to handle memory operands and you can use `emitInsBinary` then great. If not, you have to either duplicate the memory operand handling logic or get creative and create 8/16 different instructions...
* Perhaps short and long jumps should not be separate instructions, `jle` and `l_jle` is really the same instruction with 2 different encodings. This is however problematic in the case of unconditional jumps - `l_jmp` is used explicitly.
* Perhaps JCC/SETCC/CMOVCC groups should instead be a single instruction - the condition could be stored separately (where?). Then final encoding would be "ins encoding" + "condition code".
* Read/write flags don't seem to be used anywhere. Perhaps they could be put to some use in the future (validate that nothing sneaks in between CMP and JCC?) but since they are not currently used can we trust them? Delete them for now?
* The `fp` field only applies to 2 instructions - `fld` and `fstp`. Perhaps it's not worth keeping this field just for that.
* The encoding columns have fixed, global, meaning. Or so it seems. Perhaps the meaning of these columns should be instruction/format dependent. That may avoid common cases where `INS3` is used when only one encoding is actually present.

**Comment by CarolEidt** [comment]
When we undertake this refactoring, we should look at all the places where we switch on the instruction, and determine whether that is capturing an implicit property that should be explicitly captured, e.g. see #19141 where we would benefit from a property that indicates what register file(s) an instruction writes to.

**Comment by CarolEidt** [comment]
The `IsDstDstSrc` and `IsDstSrcSrc` methods are confusing and somewhat inaccurate. They are generally used to determine how to encode a 2-operand SSE instruction in a VEX encoding, but the associated flag is currently specified on a number of VEX-only instructions. It would be good to eliminate that weirdness (and perhaps, as @fiigii suggests, eliminate these methods altogether). However, AFAICT there's no real way to ensure that a supported format is being used (e.g. that we're not providing only 2 operands when 3 are required). Such a mechanism should presumably fall out of an approach that explicitly enumerates the supported formats, as is being discussed here.
